### DISCUSSION TITLE
Is there an API for AJAX, callbacks, URLs and such?

### DISCUSSION THREAD

**Discussion by salmonb** [webfx-project/webfx#15] [Q&A] [answered]
*This question was originally asked by @gnahraf, but transferred here from another discussion.*

**Comment by salmonb** [comment]
Here is what we have so far to communicate with servers: a cross-platform Web Socket API for the low level, and a cross-platform EventBus + JSON API for the high level (which is the preferred communication protocol for modern servers such as Vert.x or Spring Boot).

But I guess you just want to do a simple HTTP request? We don't presently have this API, but it shouldn't be that difficult to make. What type of data are you expecting for the result? Will you be happy with a API that just returns the result as a String? As I said, we have a cross-platform JSON API, so you can use it to parse the result if this is the format you expect. But we don't have a cross-platform XML API.

**Reply by gnahraf** [reply]
> Web Socket API for the low level, and a cross-platform EventBus + JSON API

<strike>Can you point me to the libs that implement these pls. I'm looking thru the libs (maven deps), but can't seem to find them.</strike> Found them.

> But I guess you just want to do a simple HTTP request?

Yes :)

> What type of data are you expecting for the result?

JSON. And indeed a string result would be fine.

> we have a cross-platform JSON API

I must be confused.. why is that needed? is it not possible for user code to import their own json parser? The parser I had in my mind was json-simple -- no annotation preprocessing, etc., so I thought it might work. Either way works for me, but since json parsers can be really simple code, I didn't imagine there'd be portability (cross-platform) issues. I must be missing something

**Reply by salmonb** [reply]
For JSON, just insert `dev.webfx.platform.json.Json.createObject();` somewhere in your Java code and then invoke `webfx update`. The CLI will detect that usage and automatically import our JSON API in your module, so you should be good to use it after that in your IDE. For parsing, you can use `Json.parseObject()` or `Json.parseArray()`. 

When I say our JSON API is cross-platform, I actually mean 2 things: 
1. it can be compiled by GWT. Most of other Java JSON parsers will fail to compile with GWT due to their dependency on `java.io.File`
2. it creates native JSON objects in the browser, that you can pass for example as-is to a third-party JS library.

GWT provides some JSON API, but it's designed to run in the browser only, you can't compile it for the JVM, so it's not cross-platform, whereas with our API, you can write code that can be compiled for both the browser and the JVM.

**Reply by salmonb** [reply]
> > But I guess you just want to do a simple HTTP request?

> Yes :)

So are you asking a cross-platform API to do HTTP requests in WebFX?

**Reply by gnahraf** [reply]
> Most of other Java JSON parsers will fail to compile with GWT due to their dependency on `java.io.File`

Ahh. Got it. That's good news for me. I can easily weed out `java.io.File` w/o having to change much other existing code.

> GWT .. JSON API .. designed to run in the browser only, .. not cross-platform, whereas with our API, you can write code that can be compiled for both the browser and the JVM.

Thanks for explaining what cross-platform means in this context (browser / JVM).

> So are you asking a cross-platform API to do HTTP requests in WebFX?

Yes. But honestly, I only care that it work in the browser (cross-platform). (But isn't that it be runnable on a JVM, a prerequisite for GWT to "transpile" the code to JS? Seems it'd _have to be_ cross platform)

**Comment by gnahraf** [comment]
> So are you asking a cross-platform API to do HTTP requests in WebFX?

Yes. My REST endpoints are straight HTTP/HTTPS for now (user sessions are short lived)

**Reply by salmonb** [reply]
Ok, I think I can add a cross-platform API in the WebFX Platform similar to the Web [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch), and with a Java implementation based on `java.net.http`. Anyway, this API should let you do what you want.

But it may take a couple of weeks to do, as I'm busy with other things...

**Reply by gnahraf** [reply]
> But it may take a couple of weeks to do, as I'm busy with other things

Perfect. Thank you!

**Comment by salmonb** [comment, marked as answer]
hi @gnahraf, I published the first version of the Fetch API in the WebFX Platform! Here is how to include it in your project:

1. Type `webfx bump cli` in the terminal to update your CLI to the latest version
2. Add `import dev.webfx.platform.fetch.Fetch;` in your class (or use the fully qualified class `dev.webfx.platform.fetch.Fetch` somewhere in your code)
3. Type `webfx update --clean-snapshots` (or short syntax: `webfx update -c`) at your project root directory (this new `--clean-snapshots` option should prevent the problem you had last time with old snapshots in your local Maven repository) 

 Here are a few use-cases of the Fetch API (if not yet imported, use the fully qualify class `dev.webfx.platform.console.Console`, and the `webfx update` command should import it, then you can replace the fully qualify class with an import in your IDE):

1. Fetching a text:
```java
Fetch.fetch("https://jsonplaceholder.typicode.com/posts/1")
        .onFailure(error -> Console.log("Fetch failure: " + error))
        .onSuccess(response -> {
            Console.log("Fetch success: ok = " + response.ok());
            response.text()
                    .onFailure(error -> Console.log("Text failure: " + error))
                    .onSuccess(text -> Console.log("Text Success: " + text));
        });
```

2. Fetching a JSON object:
```java
Fetch.fetch("https://jsonplaceholder.typicode.com/posts/1") // Expecting a JSON object only
        .onFailure(error -> Console.log("Fetch failure: " + error))
        .onSuccess(response -> {
            Console.log("Fetch success: ok = " + response.ok());
            response.jsonObject()
                    .onFailure(error -> Console.log("JsonObject failure: " + error))
                    .onSuccess(jsonObject -> {
                        Console.log("JsonObject Success: " + jsonObject.toJsonString());
                        Console.logNative(jsonObject); // Will display a browsable JSON object in the browser console
                    });
        });
```

3. Fetching a JSON array:
```java
Fetch.fetch("https://jsonplaceholder.typicode.com/posts") // // Expecting a JSON array only
        .onFailure(error -> Console.log("Fetch failure: " + error))
        .onSuccess(response -> {
            Console.log("Fetch success: ok = " + response.ok());
            response.jsonArray()
                    .onFailure(error -> Console.log("JsonArray failure: " + error))
                    .onSuccess(jsonArray -> {
                        Console.log("JsonArray Success: " + jsonArray.toJsonString());
                        Console.logNative(jsonArray); // Will display a browsable JSON array in the browser console
                    });
        });
```

4. Fetching a JSON object or array (if you don't know which one to expect in advance):
```java
Fetch.fetch("https://jsonplaceholder.typicode.com/posts") // Expecting a JSON object or array
        .onFailure(error -> Console.log("Fetch failure: " + error))
        .onSuccess(response -> {
            Console.log("Fetch success: ok = " + response.ok());
            response.json()
                    .onFailure(error -> Console.log("Json failure: " + error))
                    .onSuccess(json -> {
                        if (json.isObject()) {
                            JsonObject jsonObject = json.asJsonObject();
                            Console.log("JsonObject Success: " + jsonObject.toJsonString());
                            Console.logNative(jsonObject); // Will display a browsable JSON object in the browser console
                        } else if (json.isArray()) {
                            JsonArray jsonArray = json.asJsonArray();
                            Console.log("JsonArray Success: " + jsonArray.toJsonString());
                            Console.logNative(jsonArray); // Will display a browsable JSON array in the browser console
                        }
                    });
        });
```

Let me know if this works for you, or if you need another use-case.

**Comment by gnahraf** [comment]
Thank you Bruno.

I've been meaning to respond after I try it but have been busy with another
deliverable. I'll definitely try it after and loop back with you. Prolly
next week. Apologies for the delayed response.

Babak
